tiledb/api/src/column.rs

```diff
@@ -0,0 +1,416 @@
+    }
+}
+
+// Conversion for Vec or slices of primitive values for non-variadic columns.
+macro_rules! derive_primitive_as_column {
+    ($typename:ty) => {
+        impl<'data> AsColumn<'data> for &'data [$typename] {
+            fn as_column(&self) -> Column<'data> {
+                let num_bytes = self.len() * std::mem::size_of::<$typename>();
+                let data = unsafe {
+                    std::slice::from_raw_parts(
+                        self.as_ptr() as *const u8,
+                        num_bytes,
+                    )
+                };
+
+                Column::from_references(
+                    self.len() as u64,
+                    std::mem::size_of::<$typename>() as u64,
+                    data,
+                    &[0u64; 0],
+                )
+            }
+        }
+
+        impl<'data> AsColumn<'data> for Vec<$typename> {
+            fn as_column(&self) -> Column<'data> {
+                let num_bytes = self.len() * std::mem::size_of::<$typename>();
+                let data = unsafe {
+                    std::slice::from_raw_parts(
+                        self.as_ptr() as *const u8,
+                        num_bytes,
+                    )
+                };
+
+                Column::from_references(
+                    self.len() as u64,
+                    std::mem::size_of::<$typename>() as u64,
+                    data,
+                    &[0u64; 0],
+                )
+            }
+        }
+    };
+}
+
+derive_primitive_as_column!(u8);
+derive_primitive_as_column!(u16);
+derive_primitive_as_column!(u32);
+derive_primitive_as_column!(u64);
+derive_primitive_as_column!(i8);
+derive_primitive_as_column!(i16);
+derive_primitive_as_column!(i32);
+derive_primitive_as_column!(i64);
+derive_primitive_as_column!(f32);
+derive_primitive_as_column!(f64);
+
+// I couldn't figure out how to derive directly from the AsRef<str> so I've
+// made this for now. Ideally we could just implement two for Vec<AsRef<str>>
+// and &[AsRef<str>]. Rather than beat my head against it I'm moving on.
+macro_rules! derive_strings_as_column {
+    ($typename:ty) => {
+        impl AsColumn<'_> for &[$typename] {
+            fn as_column(&self) -> Column {
+                let (data, offsets) = strings_to_buffers(self);
+                Column::from_allocations(self.len() as u64, 1, data, offsets)
+            }
+        }
+
+        impl AsColumn<'_> for Vec<$typename> {
+            fn as_column(&self) -> Column {
+                let (data, offsets) = strings_to_buffers(self);
+                Column::from_allocations(self.len() as u64, 1, data, offsets)
+            }
+        }
+    };
+}
+
+derive_strings_as_column!(&str);
+derive_strings_as_column!(&String);
+derive_strings_as_column!(String);
+
+// Internal helper to convert a slice of strings to a single linear data buffer
+// and an offsets buffer.
+fn strings_to_buffers(vals: &[impl AsRef<str>]) -> (Box<[u8]>, Box<[u64]>) {
```

Comment:
This function copies all of the string data.  The alternative I intended to ask about above would be returning a `Vec<&'data [u8]>`, i.e. for each of the input strings a pointer to the data for that string, such that a copy of all the string contents is not necessary.

Reply:
Not sure I follow there. The multiple strings version requires a copy because of TileDB's APIs. The way string data is passed is like such:

```rust
// To pass something like this to TileDB
let values = vec!["foo", "bar", "baz"];

// We have to pass this:
let values = "foobarbaz";
let offsets: [u64] = [0, 3, 6];
```

There's also the obvious missing `AsColumn` implementation for tuples so that folks can do things like:

```
let values = "foobarbaz";
let offsets: [u64] = [0, 3, 6];

builder.thing((values, offsets).as_column())
```

I just haven't needed that one yet. And we'll also want to extend columns to support nullable attributes which could then look something like:

```
let values = "foobarbaz";
let offsets: [u64] = [0, 3, 3, 6];
let validity: [u8] = [1, 1, 0, 1];

builder.do_thing((values, offsets, validity).as_column())
```

And I've also suddenly realizing making offsets be u64 unconditionally means that the validity/offsets buffers won't interfere with each other which is nice.

Reply:
> The multiple strings version requires a copy because of TileDB's APIs

This was actually something I wondered about. Maybe it doesn't have to. I'll play with it on my own maybe.